metagraph/src/annotation/taxonomy/tax_classifier.hpp

```diff
@@ -0,0 +1,194 @@
+
+    /** Reads the accession version to taxid lookup table.
+    *
+    * @param [input] filepath -> a ".accession2taxid" file.
+    * @param [input] anno_matrix -> pointer to the annotation matrix
+    */
+    void read_accversion_to_taxid_map(const std::string &filepath, const graph::AnnotatedDBG *anno_matrix);
+
+    /**
+     * Update the current node_scores and best_lca by taking into account the weight of the start_node and all its ancestors.
+     *
+     * @param [input] 'start_node' -> the starting node to update 'node_scores'.
+     * @param [input] 'num_kmers_per_node[taxid]' -> the number of kmers 'k' with taxonomic_map[k]=taxid.
+     * @param [input] 'desired_number_kmers' -> the threshold score that a node has to exceed in order to be considered as a valid solution.
+     * @param [modified] 'node_scores' -> the current score for each node in the tree.
+     * @param [modified] 'nodes_already_propagated' -> the set of nodes that were previously processed.
+     * @param [modified] 'best_lca' -> the current classification prediction (node that exceeds the `desired_number_kmers` threshold and is placed as close as possible to the leaves).
+     * @param [modified] 'best_lca_dist_to_root' -> the distance to the root for the current classification prediction.
+     */
+    void update_scores_and_lca(const TaxId start_node,
+                               const tsl::hopscotch_map<TaxId, uint64_t> &num_kmers_per_node,
+                               const uint64_t desired_number_kmers,
+                               tsl::hopscotch_map<TaxId, uint64_t> *node_scores,
+                               tsl::hopscotch_set<TaxId> *nodes_already_propagated,
+                               TaxId *best_lca,
+                               uint32_t *best_lca_dist_to_root) const;
+
+    /**
+     * Get the list of LCA taxid for each kmer in a given sequences.
+     * The sequence can be given in forward or in reversed orientation.
+     */
+    virtual std::vector<TaxId> get_lca_taxids_for_seq(const std::string_view &sequence, bool reversed) const = 0;
+
+    LabelType label_type;
+
+    /**
+     * node_depth returns the depth for each node in the taxonomic tree.
+     * The root is the unique node with maximal depth and all the leaves have depth equal to 1.
+     */
+    tsl::hopscotch_map<TaxId, uint32_t> node_depth;
+
+    TaxId root_node;
+
+    /**
+     *  node_parent stores a taxonomic tree representation as a taxid to taxid parent list.
+     */
+    tsl::hopscotch_map<TaxId, TaxId> node_parent;
+
+    tsl::hopscotch_map<std::string, TaxId> accversion_to_taxid_map;
+
+    double _lca_coverage_rate;
+    double _kmers_discovery_rate;
+};
+
+class TaxonomyClsImportDB : public TaxonomyBase {
+public:
+    // todo implement
+    TaxonomyClsImportDB(const std::string &taxdb_filepath,
+                        const double lca_coverage_rate,
+                        const double kmers_discovery_rate);
+
+PRIVATE_TESTABLE:
```

Comment:
We write `private`, but then add `#define private public` in tests to access private methods. Check out how it's done in other tests.

```suggestion
  private:
```